### PR TITLE
Feature/sdk 0.45.6

### PR DESCRIPTION
**Pull request by Jedi2002**

### CODE REVIEW

**Review by gzukel — DISMISSED** [review]
+1

File: docs/tutorials/authz.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# AuthZ tutorial 
+- Authz module allows granting arbitrary privileges from one account (the granter) to another account (the grantee). Authorizations must be granted for a particular Msg service method one by one using an implementation of the Authorization interface.
+- The built in types include:`send`,`generic`,`delegate`,`unbond`,`redelegate`
+- The `generic` authorization can be used to authorize any address to execute a message on their behalf
+
+## Steps to provide authorization
+1. Grant authorization to a particular address
+```shell
+sifnoded tx authz grant sif1l7hypmqk2yc334vc6vmdwzp5sdefygj2ad93p5 generic --msg-type=/sifnode.clp.v1.MsgCreatePool --from=sif --keyring-backend=test --chain-id=localnet
```
Comment (timlind) [inline docs/tutorials/authz.md:9]: Nice!

File: x/clp/keeper/calculations_test.go [hunk 2]
```diff
@@ -438,56 +438,56 @@ func TestKeeper_CalculateWithdrawal(t *testing.T) {
...
-			name:                 "fail to convert asymmetry to Dec",
-			poolUnits:            sdk.NewUint(1),
-			nativeAssetBalance:   "1",
-			externalAssetBalance: "1",
-			lpUnits:              "1",
-			wBasisPoints:         "1",
-			asymmetry:            sdk.Int(sdk.NewUintFromString("10000000000000000000000000000000000000000000000000000000000000000000000000")),
-			panicErr:             "fail to convert 10000000000000000000000000000000000000000000000000000000000000000000000000 to cosmos.Dec: decimal out of range; bitLen: got 303, max 256",
-		},
+		//{
```
Comment (timlind) [inline x/clp/keeper/calculations_test.go:441]: Can you give a reason in the comments for commenting out these tests?

File: x/ethbridge/keeper/blacklist_test.go [hunk 3]
```diff
@@ -13,8 +16,9 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// Setting to true to use 0.45
 var (
-	isFeatureToggleSdk045 = os.Getenv("FEATURE_TOGGLE_SDK_045") != ""
+	isFeatureToggleSdk045 = os.Getenv("FEATURE_TOGGLE_SDK_045") == ""
```
Comment (timlind) [inline x/ethbridge/keeper/blacklist_test.go:21]: Can we get rid of this toggle now?

**Review by timlind — COMMENTED** [review]
Integration test aren't green

File: app/app.go [hunk 4]
```diff
@@ -117,41 +121,37 @@ var (
...
-
-				clp.AppModuleBasic{},
-				oracle.AppModuleBasic{},
-				ethbridge.AppModuleBasic{},
-				dispensation.AppModuleBasic{},
-				tokenregistry.AppModuleBasic{},
-				admin.AppModuleBasic{},
-				vesting.AppModuleBasic{},
-			},
-			FEATURE_TOGGLE_MARGIN_CLI_ALPHA_getAppModuleBasics()...,
```
Comment (canercandan) [inline app/app.go:153]: @Jedi2002 we need to keep this for Margin feature toggle to work

File: app/app.go [hunk 5]
```diff
@@ -187,8 +187,6 @@ var (
 
 type SifchainApp struct {
 	*baseapp.BaseApp
-	FEATURE_TOGGLE_MARGIN_CLI_ALPHA_SifchainApp
```
Comment (canercandan) [inline app/app.go:190]: @Jedi2002 same here if we get rid of that we are breaking with Margin additional struct required fields.

File: app/app.go [hunk 6]
```diff
@@ -728,10 +780,10 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 	paramsKeeper.Subspace(distrtypes.ModuleName)
 	paramsKeeper.Subspace(slashingtypes.ModuleName)
 	paramsKeeper.Subspace(clptypes.ModuleName)
-	FEATURE_TOGGLE_MARGIN_CLI_ALPHA_setParamsKeeper(&paramsKeeper)
 	paramsKeeper.Subspace(crisistypes.ModuleName)
 	paramsKeeper.Subspace(govtypes.ModuleName).WithKeyTable(govtypes.ParamKeyTable())
 	paramsKeeper.Subspace(ibctransfertypes.ModuleName)
 	paramsKeeper.Subspace(ibchost.ModuleName)
+	FEATURE_TOGGLE_MARGIN_CLI_ALPHA_setParamsKeeper(&paramsKeeper)
```
Comment (canercandan) [inline app/app.go:787]: @Jedi2002 is there a particular order we need to call those lines?

File: app/app_FEATURE_TOGGLE_SDK_045.go [hunk 7]
```diff
@@ -2,80 +2,3 @@
 // +build FEATURE_TOGGLE_SDK_045
 
 package app
-
```
Comment (canercandan) [inline app/app_FEATURE_TOGGLE_SDK_045.go:5]: @Jedi2002 we can actually remove this file altogether

File: app/setup_handlers.go [hunk 8]
```diff
@@ -1,34 +1,18 @@
...
-	"github.com/Sifchain/sifnode/app/ante"
 	storetypes "github.com/cosmos/cosmos-sdk/store/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	m "github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/cosmos/cosmos-sdk/x/authz"
 	"github.com/cosmos/cosmos-sdk/x/upgrade/types"
 )
 
-const releaseVersion = "0.15.0"
+const releaseVersion = "0.16.0"
```
Comment (canercandan) [inline app/setup_handlers.go:11]: @Jedi2002 I would suggest to first set a RC version such as `0.16.0-rc.2` (as we already have `0.16.0-rc.1` published with the latest PR merged to master) until we completed all the regression tests before cut the official release 0.16.0

File: cmd/sifnoded/cmd/migrate_test.go [hunk 9]
```diff
@@ -27,10 +27,6 @@ func TestMigrateGenesisDataCmd(t *testing.T) {
 	cmd.SetArgs([]string{"init", "test", "--home=" + homeDir})
 	err = svrcmd.Execute(cmd, homeDir)
 	require.NoError(t, err)
-	cmd, _ = NewRootCmd()
-	cmd.SetArgs([]string{"unsafe-reset-all"})
-	err = svrcmd.Execute(cmd, homeDir)
-	require.NoError(t, err)
```
Comment (canercandan) [inline cmd/sifnoded/cmd/migrate_test.go:33]: @Jedi2002 is there a reason we are remove those tests?

File: scripts/run.sh [hunk 10]
```diff
@@ -2,4 +2,4 @@
 
 killall sifnoded
 
-sifnoded start --trace
+sifnoded start --trace --home ~/.sifnoded/ --p2p.laddr 0.0.0.0:27655  --grpc.address 0.0.0.0:9096 --grpc-web.address 0.0.0.0:9093 --address tcp://0.0.0.0:27659 --rpc.laddr tcp://127.0.0.1:26657
```
Comment (canercandan) [inline scripts/run.sh:5]: @Jedi2002 should we use 
```
--rpc.laddr tcp://127.0.0.1:26657
```

or 

```
--rpc.laddr tcp://0.0.0.0:26657
```

considering the other params using 0.0.0.0

File: scripts/upgrade-integration.sh [hunk 11]
```diff
@@ -98,7 +97,7 @@ echo "${contents}" > $DAEMON_HOME/config/genesis.json
 
 # Add state data here if required
 
-cosmovisor start >> sifnode.log 2>&1  &
+cosmovisor start --home ~/.sifnoded/ --p2p.laddr 0.0.0.0:27655  --grpc.address 0.0.0.0:9096 --grpc-web.address 0.0.0.0:9093 --address tcp://0.0.0.0:27659 --rpc.laddr tcp://127.0.0.1:26657 >> sifnode.log 2>&1  &
```
Comment (canercandan) [inline scripts/upgrade-integration.sh:100]: same here

File: version [hunk 12]
```diff
@@ -1 +1 @@
-0.15.0
+0.16.0
```
Comment (canercandan) [inline version:1]: @Jedi2002 same question I would go with 0.16.0-rc.2 instead

File: x/clp/keeper/calculations_test.go (hunk 2 above)
Comment (canercandan) [reply x/clp/keeper/calculations_test.go:482]: @Jedi2002 I would recommend keeping the test case and increasing the number of digits of the asymmetry decimal to make it panic the same way you do with the other use cases.

File: x/ethbridge/keeper/blacklist_test.go (hunk 3 above)
Comment (canercandan) [reply x/ethbridge/keeper/blacklist_test.go:21]: correct we should get rid of this toggle as well

File: x/clp/keeper/calculations_test.go (hunk 2 above)
Comment (Jedi2002) [reply x/clp/keeper/calculations_test.go:482]: The only one that's commented not is for asymmetry which uses SDK.INT . 
The panic comes from creating the UINT which is part of the test setup, not the function it's trying to test.

if the goal is to test sdk.NewUintFromString , please let me know , we can have a separate test case for that specifically

File: app/app.go (hunk 6 above)
Comment (Jedi2002) [reply app/app.go:787]: No there isn't , I just moved the function call to end.

File: app/app_FEATURE_TOGGLE_SDK_045.go (hunk 7 above)
Comment (Jedi2002) [reply app/app_FEATURE_TOGGLE_SDK_045.go:5]: Allright  thanks, I just wasn't sure if it breaks something

File: app/setup_handlers.go (hunk 8 above)
Comment (Jedi2002) [reply app/setup_handlers.go:11]: okay

File: scripts/run.sh (hunk 10 above)
Comment (Jedi2002) [reply scripts/run.sh:5]: We can use anything tbh, I assume this is just a test script? Not used for any actual deployment

File: scripts/upgrade-integration.sh (hunk 11 above)
Comment (Jedi2002) [reply scripts/upgrade-integration.sh:100]: Again, I don't think anyone is using this script ?

File: version (hunk 12 above)
Comment (Jedi2002) [reply version:1]: okay

File: x/clp/keeper/calculations_test.go (hunk 2 above)
Comment (canercandan) [reply x/clp/keeper/calculations_test.go:482]: @Jedi2002 so are you saying those test cases you commented are not relevant anymore and if so can you explain why, thanks

File: cmd/sifnoded/cmd/migrate_test.go (hunk 9 above)
Comment (Jedi2002) [reply cmd/sifnoded/cmd/migrate_test.go:33]: The tests are still functional , just removed `unsafe-reset-all` as that command is not needed , and it has been deprecated

File: x/clp/keeper/calculations_test.go (hunk 2 above)
Comment (Jedi2002) [reply x/clp/keeper/calculations_test.go:482]: Hey, @canercandan  I think these comments are on an older commit, all other tests except the one for asymmetry. 
The asymmetry test panics at test data generation and not the function it's trying to test. We can chant about it over a call if you need more clarity.

File: app/app.go (hunk 4 above)
Comment (Jedi2002) [reply app/app.go:153]: Allright , added it back

File: app/app.go (hunk 5 above)
Comment (Jedi2002) [reply app/app.go:190]: Done

**Comment by timlind** [comment]
Not sure what happened with app.go but the margin cli tests need to be green as well.

File: app/app.go [hunk 13]
```diff
@@ -370,7 +381,16 @@ func NewSifAppWithBlacklist(
...
+	app.ClpKeeper = clpkeeper.NewKeeper(
+		appCodec,
+		keys[clptypes.StoreKey],
+		app.BankKeeper,
+		app.AccountKeeper,
+		app.TokenRegistryKeeper,
+		app.AdminKeeper,
+		app.MintKeeper,
+		app.GetSubspace(clptypes.ModuleName),
+	)
```
Comment (timlind) [inline app/app.go:393]: This needs to stay.

**Review by timlind — APPROVED** [review]
lgtm

File: app/app.go [hunk 14]
```diff
@@ -371,6 +382,16 @@ func NewSifAppWithBlacklist(
...
+	//app.ClpKeeper = clpkeeper.NewKeeper(
+	//	appCodec,
+	//	keys[clptypes.StoreKey],
+	//	app.BankKeeper,
+	//	app.AccountKeeper,
+	//	app.TokenRegistryKeeper,
+	//	app.AdminKeeper,
+	//	app.MintKeeper,
+	//	app.GetSubspace(clptypes.ModuleName),
+	//)
```
Comment (canercandan) [inline app/app.go:394]: @Jedi2002 can remove this commented block of code?

File: x/ethbridge/keeper/blacklist.go [hunk 15]
```diff
@@ -64,9 +65,14 @@ func (k Keeper) GetBlacklist(ctx sdk.Context) []string {
 	defer iter.Close()
 	for ; iter.Valid(); iter.Next() {
 		key := iter.Key()
+		fmt.Println(key)
```
Comment (canercandan) [inline x/ethbridge/keeper/blacklist.go:68]: @Jedi2002 can we remove this print line?

**Review by canercandan — APPROVED** [review]
LGTM with minor comments

**Pull request merged**